src/main/java/edu/byu/cs/autograder/compile/ModuleIndependenceVerifier.java

```diff
@@ -0,0 +1,101 @@
+        try {
+            Set<File> serverFiles = reader.filesMatching(".*server/src/main/java/.*").collect(Collectors.toSet());
+            Set<String> serverPackages = packageNames(serverFiles, reader);
+
+            Set<File> clientFiles = reader.filesMatching(".*client/src/main/java/.*").collect(Collectors.toSet());
+            Set<String> clientPackages = packageNames(clientFiles, reader);
+
+            removeCommonItems(serverPackages, clientPackages);
+
+            checkImports(context, reader, serverFiles, clientPackages);
+            checkImports(context, reader, clientFiles, serverPackages);
+        } catch (IOException e) {
+            throw new GradingException("Unable to verify module independence", e);
+        }
+    }
+
+    private Set<String> packageNames(Set<File> files, StudentCodeReader reader) throws IOException {
+        Set<String> packages = new HashSet<>();
+        for(File file : files) {
+            if(!file.getName().endsWith(".java")) continue;
+            List<String> contents = reader.getFileContents(file);
+            if(contents.isEmpty()) continue;
+            String firstLine = contents.getFirst();
+            if(firstLine.startsWith("package")) {
+                packages.add(firstLine.substring(8, firstLine.indexOf(';')));
+            }
+        }
+        return packages;
+    }
+
+    private void checkImports(GradingContext context, StudentCodeReader reader, Set<File> files, Set<String> packages)
+            throws IOException {
+        for(File file : files) {
+            List<String> contents = reader.getFileContents(file);
+            for (int i = 0; i < contents.size(); i++) {
+                String line = contents.get(i).trim();
+                if (line.isEmpty()) continue;
+                if (line.matches("^(?!import|package|//)(/\\*\\*|@|.*\\b(class|record|enum|@?interface)\\b).*$")) break;
+
+                String packageImport = getPackageImport(line);
+                if (packageImport != null && packages.contains(packageImport)) {
+                    String warning = ("File %s imports from package %s (line %d), which exists in another module. " +
+                            "The client and server modules should be independent")
+                            .formatted(context.stageRepo().toPath().relativize(file.toPath()), packageImport, i + 1);
+                    context.observer().notifyWarning(warning);
+                }
+            }
+        }
+    }
+
+    /**
+     * Gets just the package being imported from a java import statement
+     * <p>For example, "import java.util.Set;" would return "java.util"
+     * <p>For example, "import static java.sql.Types.NULL;" would return "java.sql"
+     *
+     * @param line java import statement
+     * @return package name
+     */
+    private static String getPackageImport(String line) {
+        int firstSpace = line.indexOf(' ');
```

Comment:
Per a discussion with @19mdavenport, let's use [this REGEX](https://regex101.com/r/rSSd12) to verify that the line is a valid import line. If it's not, we'll `return null`, signifying that this shouldn't be searched.

This additional check is intended to be more precise than the basic but fast check for `line.startsWith("import")` that is called before evaluating this method. Honestly, it probably makes sense to put those two statements next to each other for clarity, but that decision is left up to the implementer :)